### PR TITLE
network: Handle routes with empty/nil destination as "default" destination

### PR DESCRIPTION
**Pull request by sboeuf**
Until now, we were ignoring routes that we scanned in case the destination was nil, but we were missing the most important route, i.e the one redirecting everything else to the docker gateway (usually 172.17.0.1).

### CODE REVIEW

File: hyperstart.go [hunk 1]
```diff
@@ -172,18 +172,23 @@ func (h *hyper) buildNetworkInterfacesAndRoutes(pod Pod) ([]hyperstart.NetworkIf
...
 				gateway = ""
 			}
 
+			destination := r.Dst.String()
+			if destination == emptyRouteDest {
+				destination = defaultRouteDest
+			}
+
+			// Skip IPv6 because not supported by hyperstart
+			if destination != defaultRouteDest && r.Dst.IP.To4() == nil {
```
Comment (jodh-intel) [inline hyperstart.go]: Is it worth adding a log call here so we are constantly reminded about this limitation? Is there an issue open to document that limitation in fact?

Comment (sboeuf) [inline hyperstart.go]: @jodh-intel I don't think we have any issues opened for that. Let me add a comment.

Comment (sboeuf) [inline hyperstart.go]: Updated!

File: hyperstart.go [hunk 2]
```diff
@@ -172,18 +172,24 @@ func (h *hyper) buildNetworkInterfacesAndRoutes(pod Pod) ([]hyperstart.NetworkIf
...
-			}
-
 			gateway := r.GW.String()
 			if gateway == "<nil>" {
 				gateway = ""
 			}
 
+			destination := r.Dst.String()
+			if destination == emptyRouteDest {
+				destination = defaultRouteDest
```
Comment (sboeuf) [inline hyperstart.go]: I have created a subfunction so that I could unit test it easily.

File: network.go [hunk 3]
```diff
@@ -30,6 +30,11 @@ import (
 	"golang.org/x/sys/unix"
 )
 
+// Introduces constants related to network routes.
+const (
+	emptyRouteDest = "0.0.0.0/0"
+)
```
Comment (sameo) [inline network.go]: `0.0.0.0` is the IP for the default route, so let's call it this way :)

File: cnm.go [hunk 4]
```diff
@@ -95,7 +95,12 @@ func (n *cnm) createResult(iface net.Interface, addrs []net.Addr, routes []netli
 	var resultRoutes []*cniTypes.Route
 	for _, route := range routes {
 		if route.Dst == nil {
-			continue
+			_, ipNet, err := net.ParseCIDR(emptyRouteDest)
```
Comment (sameo) [inline cnm.go]: `s/ipNet/defaultRoute`

File: network.go (hunk 3 above)
Comment (sameo) [inline network.go]: And `defaultRouteDest` could be call `defaultRouteLabel`

Comment (sboeuf) [inline network.go]: Done

File: cnm.go (hunk 4 above)
Comment (sboeuf) [inline cnm.go]: Done

**Comment by sboeuf** [comment]
PR updated.

**Comment by sameo** [comment]
LGTM

**Comment by sboeuf** [comment]
Should we wait for @mcastelino on this one ? I think he's on vacation.

**Comment by sameo** [comment]
> Should we wait for @mcastelino on this one ? I think he's on vacation.

Yes, please.

**Comment by mcastelino** [comment]
LGTM

**Pull request merged**